io/src/real_sense_2_grabber.cpp

```diff
@@ -0,0 +1,394 @@
+
+    auto sp = points.get_profile ().as<rs2::video_stream_profile> ();
+    cloud->width = sp.width ();
+    cloud->height = sp.height ();
+    cloud->is_dense = false;
+    cloud->points.resize (points.size ());
+
+    auto vertices_ptr = points.get_vertices ();
+    auto texture_ptr = points.get_texture_coordinates ();
+
+#ifdef _OPENMP
+#pragma omp parallel for 
+#endif
+    for (int index = 0; index < cloud->points.size (); ++index)
+    {
+      auto ptr = vertices_ptr + index;
+      auto uvptr = texture_ptr + index;
+      auto& p = cloud->points[index];
+
+      p.x = ptr->x;
+      p.y = ptr->y;
+      p.z = ptr->z;
+
+      std::tie (p.r, p.g, p.b) = getTextureColor (rgb, uvptr->u, uvptr->v);
+
+      p.a = 255;
+    }
+
+    return cloud;
+  }
+
+  pcl::PointCloud<pcl::PointXYZRGBA>::Ptr 
+  RealSense2Grabber::convertRGBADepthToPointXYZRGBA (const rs2::points & points, rs2::video_frame & rgb)
+  {
+    pcl::PointCloud<pcl::PointXYZRGBA>::Ptr cloud (new pcl::PointCloud<pcl::PointXYZRGBA> ());
+
+    auto sp = points.get_profile ().as<rs2::video_stream_profile> ();
+    cloud->width = sp.width ();
+    cloud->height = sp.height ();
+    cloud->is_dense = false;
+    cloud->points.resize (points.size ());
+
+    auto vertices_ptr = points.get_vertices ();
+    auto texture_ptr = points.get_texture_coordinates ();
+
+#ifdef _OPENMP
+#pragma omp parallel for 
+#endif
+    for (int index = 0; index < cloud->points.size (); ++index)
+    {
+      auto ptr = vertices_ptr + index;
+      auto uvptr = texture_ptr + index;
+      auto& p = cloud->points[index];
+
+      p.x = ptr->x;
+      p.y = ptr->y;
+      p.z = ptr->z;
+
+      std::tie (p.r, p.g, p.b) = getTextureColor (rgb, uvptr->u, uvptr->v);
+
+      p.a = 255;
+    }
+
+    return cloud;
+  }
+
+  std::tuple<uint8_t, uint8_t, uint8_t> 
+  RealSense2Grabber::getTextureColor (rs2::video_frame &texture, float u, float v)
+  {
+    auto ptr = dynamic_cast<rs2::video_frame*>(&texture);
+
+    const int w = ptr->get_width (), h = ptr->get_height ();
+    int x = std::min (std::max (int (u*w + .5f), 0), w - 1);
+    int y = std::min (std::max (int (v*h + .5f), 0), h - 1);
+    int idx = x * ptr->get_bytes_per_pixel() + y * ptr->get_stride_in_bytes ();
+    const auto texture_data = reinterpret_cast<const uint8_t*>(ptr->get_data ());
+    return std::make_tuple (texture_data[idx], texture_data[idx + 1], texture_data[idx + 2]);
+  }
+
+  uint8_t
+  RealSense2Grabber::getTextureIntensity(rs2::video_frame &texture, float u, float v)
+  {
+    auto ptr = dynamic_cast<rs2::video_frame*>(&texture);
+
+    const int w = ptr->get_width (), h = ptr->get_height ();
+    int x = std::min (std::max (int (u*w + .5f), 0), w - 1);
+    int y = std::min (std::max (int (v*h + .5f), 0), h - 1);
+    int idx = x * ptr->get_bytes_per_pixel () + y * ptr->get_stride_in_bytes ();
+    const auto texture_data = reinterpret_cast<const uint8_t*>(ptr->get_data ());
+    return texture_data[idx];
+  }
```

Comment:
The two methods above look way too similar and probably can be refactored to minimize code repetition.

Reply:
I honestly can't see a nice way to do this as the cloud pointers I'm declaring are two distinct structures.  pcl::PointXYZRGB and pcl::PointXYZRGBA.  Do you have an idea of how it could work better?

Reply:
Make this method templated on the point type?